docs/spec/lightclient/Lightclient_A_1.tla

```diff
@@ -0,0 +1,463 @@
+       IN
+        \* initially, fetchedLightBlocks is a function of one element, i.e., TRUSTED_HEIGHT
+        /\ fetchedLightBlocks = [h \in {TRUSTED_HEIGHT} |-> trustedLightBlock]
+        \* initially, lightBlockStatus is a function of one element, i.e., TRUSTED_HEIGHT
+        /\ lightBlockStatus = [h \in {TRUSTED_HEIGHT} |-> "StateVerified"]
+        \* the latest verified block the the trusted block
+        /\ latestVerified = trustedLightBlock
+
+\* copy the block from the light store        
+LightStoreGetInto(block, height) ==
+    block = fetchedLightBlocks[height]
+
+\* block should contain a copy of the block from the reference chain, with a matching commit
+CopyLightBlockFromChain(block, height) ==
+    LET ref == blockchain[height]
+        lastCommit ==
+          IF height < ULTIMATE_HEIGHT
+          THEN blockchain[height + 1].lastCommit
+            \* for the ultimate block, which we never use, as ULTIMATE_HEIGHT = TARGET_HEIGHT + 1
+          ELSE blockchain[height].VS 
+    IN
+    block = [header |-> ref, Commits |-> lastCommit]      
+
+\* Either the primary is correct and the block comes from the reference chain,
+\* or the block is produced by a faulty primary.
+\*
+\* TODO: add a traceability tag when Josef adds it in the English spec
+FetchLightBlockInto(block, height) ==
+    IF IS_PRIMARY_CORRECT
+    THEN CopyLightBlockFromChain(block, height)
+    ELSE BC!IsLightBlockAllowedByDigitalSignatures(height, block)
+
+\* add a block into the light store    
+\*
+\* TODO: add a traceability tag when Josef adds it in the English spec
+LightStoreUpdateBlocks(lightBlocks, block) ==
+    LET ht == block.header.height IN    
+    [h \in DOMAIN lightBlocks \union {ht} |->
+        IF h = ht THEN block ELSE lightBlocks[h]]
+
+\* update the state of a light block      
+\*
+\* TODO: add a traceability tag when Josef adds it in the English spec
+LightStoreUpdateStates(statuses, ht, blockState) ==
+    [h \in DOMAIN statuses \union {ht} |->
+        IF h = ht THEN blockState ELSE statuses[h]]      
+
+\* Check, whether newHeight is a possible next height for the light client.
+\*
+\* TODO: add a traceability tag when Josef adds it in the English spec
+ScheduleTo(newHeight, pNextHeight, pTargetHeight, pLatestVerified) ==
+    LET ht == pLatestVerified.header.height IN
+    \/ /\ ht = pNextHeight
+       /\ ht < pTargetHeight
+       /\ pNextHeight < newHeight
+       /\ newHeight <= pTargetHeight
+    \/ /\ ht < pNextHeight
+       /\ ht < pTargetHeight
+       /\ ht < newHeight
+       /\ newHeight < pNextHeight
+    \/ /\ ht = pTargetHeight
+       /\ newHeight = pTargetHeight
+
+\* The loop of VerifyToTarget.
+\*
+\* TODO: add a traceability tag when Josef adds it in the English spec
+VerifyToTargetLoop ==
+      \* the loop condition is true
+    /\ latestVerified.header.height < TARGET_HEIGHT
+      \* pick a light block, which will be constrained later
+    /\ \E current \in BC!LightBlocks:
```

Comment:
This is a bit confusing. We pick any block from the blockchain only to immediately overwrite it with the block with `nextHeight` either from the light store or from chain. Can we remove this line altogether?

Reply:
> We pick any block from the blockchain only to immediately overwrite

Here is the place where TLA+ differs from a programming language. In a model checker like TLC, all blocks will be enumerated and then the blocks that do not match the constraint will be rejected. (No overwriting happens.) In Apalache, this will be just a constraint, the solver has to come up with a value of `current` that matches the constraint.

What we do here is a common idiom in verification: Non-deterministically guess a value and then constrain it with your assumptions. Essentially, all symbolic tools use this idiom, one way or another.